service/matching/physical_task_queue_key.go

```diff
@@ -174,36 +195,102 @@ func ParsePhysicalTaskQueueKey(persistenceName string, namespaceId string, taskT
+//nolint:revive // cognitive complexity will be simpler once versioning 1-2 is cleaned up
+func parseSuffix(persistenceName string, suffix string) (partition int, versionSet string, buildId string, deploymentName string, err error) {
+	if buildIdOff := strings.LastIndex(suffix, deploymentNameDelimiter); buildIdOff == 0 {
+		return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
+	} else if buildIdOff > 0 {
+		deploymentNameBytes, err := base64.URLEncoding.DecodeString(suffix[:buildIdOff])
+		if err != nil {
+			return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
+		}
+		deploymentName = string(deploymentNameBytes)
+
+		suffix = suffix[buildIdOff+1:]
+		if partitionOff := strings.LastIndex(suffix, buildIdDelimiter); partitionOff == 0 {
+			return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
+		} else if partitionOff > 0 {
+			buildIdBytes, err := base64.URLEncoding.DecodeString(suffix[:partitionOff])
+			if err != nil {
+				return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
+			}
+			buildId = string(buildIdBytes)
+			suffix = suffix[partitionOff+1:]
+		}
+	} else if partitionOff := strings.LastIndex(suffix, buildIdDelimiter); partitionOff == 0 {
+		return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
 	} else if partitionOff > 0 {
 		buildIdBytes, err := base64.URLEncoding.DecodeString(suffix[:partitionOff])
 		if err != nil {
-			return 0, "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
+			return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
 		}
 		buildId = string(buildIdBytes)
 		suffix = suffix[partitionOff+1:]
 	} else if partitionOff := strings.LastIndex(suffix, versionSetDelimiter); partitionOff == 0 {
-		return 0, "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
+		return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
 	} else if partitionOff > 0 {
 		// pull out version set
 		versionSet, suffix = suffix[:partitionOff], suffix[partitionOff+1:]
 	}
 
 	partition, err = strconv.Atoi(suffix)
 	if err != nil || partition < 0 || (partition == 0 && len(versionSet) == 0 && len(buildId) == 0) {
-		return 0, "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
+		return 0, "", "", "", fmt.Errorf("%w: %s", ErrInvalidPersistenceName, persistenceName)
 	}
-	return partition, versionSet, buildId, err
+	return partition, versionSet, buildId, deploymentName, err
 }
 
 func (q *PhysicalTaskQueueKey) IsVersioned() bool {
-	return q.versionSet != "" || q.buildId != ""
+	return q.version.IsVersioned()
+}
+
+// Version returns a pointer to the physical queue version key. Caller must not manipulate the
+// returned value.
+func (q *PhysicalTaskQueueKey) Version() *PhysicalTaskQueueVersion {
```

Comment:
How about we just return it by value? I imagine most uses of it are going to get inlined away anyway so by value seem cleaner